### ISSUE TITLE
Import imaging modules

### ISSUE THREAD

**Issue by hewgreen** [HumanCellAtlas/ingest-central#219]
**What problem does the suggested enhancement solve? Please describe.**

[channels.json](https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/module/protocol/channel.json) and [imaging_targets.json](https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/module/protocol/imaging_target.json) are protocol modules imported by [imaging_protocol.json](https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/type/protocol/imaging/imaging_protocol.json). Similarly to contacts and publications we would like to import these as separate tabs in the spreadsheet as they are arrays.

We only want a temporary fix ASAP similar to the previous solution. A general solution should be scheduled later on.

**What type of enhancement is this?**

usability

**How will this enhancement benefit wrangers and/or end-users?**
This ticket aims to unblock Zina and Matt so they can attempt an ingestion of a small example imaging dataset. We may uncover other blockers that need more work but these can be estimated and delivered at a later date.
**How complex to implement do you estimate this enhancement will be? (High, Medium, Low)**
very low
**How much benefit do you estimate this enhancement will provide? (High, Medium, Low)**
high
**How urgent is this and is there a specific date it needs to be done by? (High, Medium, Low)**
high urgency
**Describe your preferred solution**

As discussed with Simon and Alegria a short term solution is good.

**Comment by aaclan-ebi** [comment]
On second look, the fix may not be as simple as it seems as the imaging protocol modules aren't exactly similar with project modules which is associated to only one project. The user would need to provide an imaging protocol id column in the module worksheet and in the importer, we need to implement the logic to map each module rows to specific imaging protocol row.

**Comment by hewgreen** [comment]
@simonjupp Could we have a chat about the best long and short term options for this problem? Essentially channels and targets are arrays inside the protocol. However, these will be used on every bundle rather than once per project as Alegria highlighted. We should also consider reusability. You may want to reuse these arrays within different protocols although I predict there will only be one per submission. Maybe they would be best as protocol types rather than modules.

**Comment by hewgreen** [comment]
Decisions for this implementation are as follows:

1. targets-channel links will be text only and not officially recognised by ingest. Some renaming is needed. 'channel_id' will be used to make this link. Matt and Zina to make these changes.
2. Even as a minimum we need a general mechanism. Tabs should use dot notation in the tab name so that we can implement arrays.
3. We need to support multiple channel/target arrays in a project but one per imaging protocol. For this we need to use the previous protocol ID field to identify which channel/target belongs to which protocol.

Next steps:
@zperova is wrangling an example for Alegria to look at.
@aaclan-ebi is deciding how much work this will be to implement.
@morrisonnorman will prioritise this work.

**Comment by justincc** [comment]
Estimation will be joint with all the devs in one of our pre-estimation planning sessions.

**Comment by zperova** [comment]
@justincc Why not for GA? I think we should aim to have imaging data in for the next F2F.

**Comment by lauraclarke** [comment]
@zperova We need to consider developer capacity and priorities. There are probably intermediate steps that you and Matt can take to ensure there is exemplar data before the F2F but full ingestion isn't necessarily something we have the capacity for. We should talk through what is needed more fully and figure out the best strategy

**Comment by lauraclarke** [comment]
Lets sit down and talk this through so we have a clear plan

**Comment by justincc** [comment]
I'm anticipating we'll discuss in the imaging meeting next Monday.

**Comment by hewgreen** [comment]
- [ ] @zperova to hand roll an example of the JSON to inform the requirement

Currently on hold waiting for data to be submitted.

**Comment by justincc** [comment]
We now have potentially 2 imaging datasets for GA. Need to discuss this in management group. @alegria did initial investigation, implementation _might_ be relatively simple but need to confirm this. @norman will talk with @laura and @tbudett about prioritization. Will not do in 2019-01-08 sprint.

**Comment by hewgreen** [comment]
NB targets.json (currently a separate module) is an array that imports another module purchased_reagents.json (not an array). This would need to be supported for this current schema to work. @rdgoite @justincc

**Comment by justincc** [comment]
Array tab support being implemented in #203

**Comment by rdgoite** [comment]
For now, the scope of the changes we're making only involves up to the first degree of nested array of objects. Once we've implemented this, we can look more into wheter we can fit the spreadsheet to work in this configuration or decide to go with further degrees of array nesting.

The case of `purchased_reagents` seems to be a little different as it doesn't involve array nesting. I think this is currently supported by our importer/broker framework so we should be ok in this area (once we have the module tabs implemented).

**Comment by hewgreen** [comment]
its a nested array (also a module) that imports another module (not an array)

imaging_protocol -> targets -> purchased reagents

**Comment by justincc** [comment]
Imaging modules supported. Epic for all other imaging work (e.g. validation) is #340
